### PR TITLE
[1.20.2] Fix BER render bounding boxes of various blocks

### LINKED ISSUE #177: When a book is placed on a lectern, it can disappear depending on the viewing angle. (Culling bug?)
**Minecraft Version:** 1.20.1

**NeoForge Version:** 47.1.79

Vanilla:

Neo:

Video:

https://github.com/neoforged/NeoForge/assets/68275908/9e01488a-430a-4544-be53-5a407bdec9a9

### PR DESCRIPTION
**Pull request by XFactHD**
This PR adds an explicit render bounding box for the Lectern to fix the book being incorrectly culled at certain viewing angles. This is needed because the collision box, which the render BB is based by default on, is only 15 pixels high while the highest point of the book model is at approximately 19.5 pixels. The value being used for the height of the BB is empirically determined by placing two blocks next to the lectern and literally counting pixels.

Fixes #177

### CODE REVIEW

**Comment by Technici4n** [comment]
I think that we need to rethink the hook a bit. We don't want to be hardcoding vanilla bounding boxes, IMO.

**Comment by XFactHD** [comment]
We can't not hardcode them, the whole point in those few cases is that the render BB has to be different than any other bounding box the block specifies. We can cut down the cases a bit though as the chests for example shouldn't need any special BBs anymore AFAICT.

I do however think that a rework in terms of discoverability is in order as throwing all cases into `IForgeBlockEntity` sucks in that regard. Moving the special cases into the individual BE classes (i.e. via mixins once those are usable in Forge itself) would improve that significantly and would also have the nice benefit that the BB can be cached instead of having to recompute it on every call

**Comment by XFactHD** [comment]
Due to the mention of the dragon head suffering from the same issue, I re-checked the culling behaviour of all BEs with BERs and found that the render bounding boxes of all skull blocks, the bell, the two campfires, the ender chest and the enchanting table also have this issue. These are now also fixed and the lectern case got a slightly different height value to also be usable for the enchanting table.

File: src/main/java/net/neoforged/neoforge/common/extensions/IBlockEntityExtension.java [hunk 1]
```diff
@@ -103,13 +104,15 @@ default AABB getRenderBoundingBox() {
...
-        if (block == Blocks.ENCHANTING_TABLE) {
-            bb = new AABB(pos, pos.offset(1, 1, 1));
-        } else if (block == Blocks.CHEST || block == Blocks.TRAPPED_CHEST) {
+        if (block == Blocks.DRAGON_HEAD) {
+            bb = new AABB(pos.getX() - .5, pos.getY(), pos.getZ() - .5, pos.getX() + 1.5, pos.getY() + 1.5, pos.getZ() + 1.5);
+        } else if (block == Blocks.BELL || block == Blocks.CAMPFIRE || block == Blocks.SOUL_CAMPFIRE || block instanceof SkullBlock) {
+            bb = new AABB(pos);
+        } else if (block == Blocks.CHEST || block == Blocks.TRAPPED_CHEST || block == Blocks.ENDER_CHEST) {
             bb = new AABB(pos.offset(-1, 0, -1), pos.offset(2, 2, 2));
-        } else if (block == Blocks.STRUCTURE_BLOCK) {
```
Comment (marchermans) [inline src/main/java/net/neoforged/neoforge/common/extensions/IBlockEntityExtension.java:110]: This seems a bit weird to me....? Is it GitHub that is stupid or did you remove the infinite bounds on the structure block?

Comment (sciwhiz12) [reply src/main/java/net/neoforged/neoforge/common/extensions/IBlockEntityExtension.java:110]: The initial value of `bb` is the infinite extent AABB, and the last condition in this `if`-`else` chain now checks that the block is _not_ the structure block. Just a small cleanup to reduce a bit of unnecessary duplication.

**Comment by Technici4n** [comment]
IMO we need to reconsider this hook. Haven't had a chance to give it a closer look since last week unfortunately

File: src/main/java/net/neoforged/neoforge/client/extensions/IBlockEntityRendererExtension.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+package net.neoforged.neoforge.client.extensions;
+
+import net.minecraft.client.renderer.blockentity.BlockEntityRenderer;
+import net.minecraft.world.level.block.entity.BlockEntity;
+import net.minecraft.world.phys.AABB;
+
+public interface IBlockEntityRendererExtension<T extends BlockEntity> {
+    AABB INFINITE_EXTENT_AABB = new net.minecraft.world.phys.AABB(Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY, Double.POSITIVE_INFINITY);
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/client/extensions/IBlockEntityRendererExtension.java]: You don't need the full name qualification here :P

Could you also add a small comment like there was before?

**Review by embeddedt — DISMISSED** [review]

File: src/main/java/net/neoforged/neoforge/client/extensions/IBlockEntityRendererExtension.java (hunk 2 above)
Comment (XFactHD) [reply src/main/java/net/neoforged/neoforge/client/extensions/IBlockEntityRendererExtension.java]: Blame the copy-paste :D
Comment added.

**Review by embeddedt — APPROVED** [review]

**Pull request merged**